src/main/scala/chiseltest/formal/backends/smt/SMTLibResponseParser.scala

```diff
@@ -72,33 +72,56 @@ private object SMTLibResponseParser {
   }
 }
 
-sealed trait SExpr
+sealed trait SExpr {
+  def isEmpty: Boolean
+}
 case class SExprNode(children: List[SExpr]) extends SExpr {
   override def toString = children.mkString("(", " ", ")")
+  override def isEmpty: Boolean = children.isEmpty || children.forall(_.isEmpty)
 }
 case class SExprLeaf(value: String) extends SExpr {
   override def toString = value
+  override def isEmpty: Boolean = value.trim.isEmpty
 }
 
 /** simple S-Expression parser to make sense of SMTLib solver output */
 object SExprParser {
   def parse(line: String): SExpr = {
-    val tokens = line
-      .replace("(", " ( ")
-      .replace(")", " ) ")
-      .split("\\s+")
-      .filterNot(_.isEmpty)
-      .toList
+    val tokens = tokenize(line)
 
-    assert(tokens.nonEmpty)
-    if (tokens.head == "(") {
+    if (tokens.isEmpty) {
+      SExprLeaf("")
+    } else if (tokens.head == "(") {
       parseSExpr(tokens.tail)._1
     } else {
-      assert(tokens.tail.isEmpty)
+      assert(tokens.tail.isEmpty, s"multiple tokens, but not starting with a `(`:\n${tokens.mkString(" ")}")
       SExprLeaf(tokens.head)
     }
   }
 
+  // tokenization with | as escape character
+  private def tokenize(line: String): List[String] = {
+    var tokens: List[String] = List()
+    var inEscape = false
+    var tmp = ""
+    def finish(): Unit = {
+      if (tmp.nonEmpty) {
+        tokens = tokens :+ tmp
+        tmp = ""
+      }
+    }
+    line.foreach {
+      case '('                                   => finish(); tokens = tokens :+ "("
+      case ')'                                   => finish(); tokens = tokens :+ ")"
+      case '|' if inEscape                       => tmp += '|'; finish(); inEscape = false
+      case '|' if !inEscape                      => finish(); inEscape = true; tmp = "|"
+      case ' ' | '\t' | '\r' | '\n' if !inEscape => finish()
+      case other                                 => tmp += other
+    }
+    finish()
+    tokens
+  }
```

Comment:
```suggestion
  private def tokenize(line: String): List[String] = {
    val tokens = mutable.ListBuffer.empty[String]
    var inEscape = false
    var tmp = ""
    def finish(): Unit = {
      if (tmp.nonEmpty) {
       tokens += tmp
        tmp = ""
      }
    }
    line.foreach {
      case '('                                   => finish(); tokens += "("
      case ')'                                   => finish(); tokens += ")"
      case '|' if inEscape                       => tmp += '|'; finish(); inEscape = false
      case '|' if !inEscape                      => finish(); inEscape = true; tmp = "|"
      case ' ' | '\t' | '\r' | '\n' if !inEscape => finish()
      case other                                 => tmp += other
    }
    finish()
    tokens.toList
  }
```
Appending to a `List` is going to be pretty slow for large output Lists (n^2). If you're really committed to using a `List`, you'll get much better performance this way. I'd also consider doing a similar thing with a `VectorBuilder` for more efficient memory use.

Reply:
This is a great suggestion! I will change it.